File: androidApp/src/main/java/com/medhelp/medhelp/ui/doctor/DoctorsPresenter.java
package com.medhelp.medhelp.ui.doctor;


import android.content.Context;
import com.medhelp.medhelp.data.model.AllDoctorsResponse;
import com.medhelp.medhelp.data.model.SpecialtyList;
import com.medhelp.medhelp.data.network.NetworkManager;
import com.medhelp.medhelp.data.pref.PreferencesManager;
import com.medhelp.medhelp.utils.timber_log.LoggingTree;
import com.medhelp.shared.model.UserResponse;

import java.util.ArrayList;
import java.util.List;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

public class DoctorsPresenter implements DoctorsHelper.Presenter
{
    private List<AllDoctorsResponse> allDoc;
    Context context;
    DoctorsHelper.View view;

    PreferencesManager prefManager;
    NetworkManager networkManager;

    public DoctorsPresenter(Context context, DoctorsHelper.View view)
    {
        this.context=context;
        this.view=view;

        prefManager = new PreferencesManager(context);
        networkManager = new NetworkManager(prefManager);

        //int tt = prefManager.getCurrentUserInfo().getIdCenter();
    }

    @Override
    public void removePassword()
    {
        prefManager.setCurrentPassword("");
        prefManager.setUsersLogin(null);
    }

    @Override
    public void getDoctorList(int idSpec)
    {

        if(idSpec==-1  && allDoc!=null)
        {
            view.updateView(allDoc);
            view.hideLoading();
            return;
        }

        if(allDoc!=null)
        {
            view.updateView(sortAllDoc(idSpec));
            view.hideLoading();
            return;
        }

        view.showLoading();

        CompositeDisposable cd = new CompositeDisposable();
        cd.add(networkManager
                .getAllDoctors()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(response ->
                        {
                            if (response.getResponses().size() <= 0) {
                                view.updateView(null);
                            } else {
                                allDoc = response.getResponses();

                                if(view!=null)
                                    view.updateView(response.getResponses());
                            }
                            view.hideLoading();
                            cd.dispose();
                        },
                        throwable ->
                        {
                            Timber.tag("my").e(LoggingTree.getMessageForError(throwable,"DoctorsPresenter$getDoctorList "));

                            view.hideLoading();
                            view.showErrorScreen();
                            cd.dispose();
                        })
        );


    }

    private List<AllDoctorsResponse> sortAllDoc(int idSpec)
    {
        List<AllDoctorsResponse> sortList =new ArrayList<>();

        for(int i=0;i<allDoc.size();i++)
        {
            AllDoctorsResponse doc=allDoc.get(i);

            if(doc.getId_specialties_int_list()==null)
                continue;

            if(doc.getId_specialties_int_list().size()==1  && doc.getId_specialties_int_list().get(0)==idSpec)
            {
                sortList.add(doc);
                continue;
            }

            for (int j=0;j<doc.getId_specialties_int_list().size();j++)
            {
                if(doc.getId_specialties_int_list().get(j)==idSpec)
                {
                    sortList.add(doc);
                }
            }
        }

        return sortList;
    }

    @Override
    public void getSpecialtyByCenter()
    {
        view.showLoading();

        CompositeDisposable cd = new CompositeDisposable();
        cd.add(networkManager
                .getCategoryApiCall()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(SpecialtyList::getSpec)
                .subscribe(response ->
                {
                    try
                    {
                        view.updateSpecialty(response);
                    } catch (Exception e)
                    {
                        Timber.tag("my").e(LoggingTree.getMessageForError(e,"DoctorsPresenter$getSpecialtyByCenter$1 "));
                    }

                    cd.dispose();
                }, throwable ->
                {
                    view.hideLoading();
                    Timber.tag("my").e(LoggingTree.getMessageForError(throwable,"DoctorsPresenter$getSpecialtyByCenter$2 "));
                    view.showErrorScreen();
                    cd.dispose();
                }));
    }

    @Override
    public void unSubscribe()
    {
        view.hideLoading();
    }

    @Override
    public String getUserToken() {
        return prefManager.getCurrentUserInfo().getApiKey();
    }

    @Override
    public UserResponse getCurrentUser()
    {
        return prefManager.getCurrentUserInfo();
    }

    @Override
    public void setCurrentUser(UserResponse user)
    {
        prefManager.setCurrentUserInfo(user);
    }
}

File: androidApp/src/main/java/com/medhelp/medhelp/ui/doctor/DoctorsPresenter.kt
package com.medhelp.medhelp.ui.doctor

import com.medhelp.medhelp.utils.timber_log.LoggingTree.Companion.getMessageForError
import com.medhelp.medhelp.data.pref.PreferencesManager
import com.medhelp.newmedhelp.model.AllDoctorsResponse
import com.medhelp.shared.model.UserResponse
import com.medhelp.shared.network.NetworkManager
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import timber.log.Timber
import java.util.ArrayList

class DoctorsPresenter(val view: DoctorsFragment) {
    private var allDoc: List<AllDoctorsResponse>? = null
    var prefManager: PreferencesManager
    var networkManager: NetworkManager

    val mainScope = MainScope()

    init {
        prefManager = PreferencesManager(view.requireContext())
        networkManager = NetworkManager()
    }

    fun removePassword() {
        prefManager.currentPassword = ""
        prefManager.usersLogin = null
    }

    fun getSpecialtyByCenter() {
        view!!.showLoading()

        mainScope.launch {
            kotlin.runCatching {
                networkManager.getCategoryApiCall(prefManager.currentUserInfo!!.apiKey!!, prefManager.centerInfo!!.db_name!!,prefManager.currentUserInfo!!.idUser.toString(),prefManager.currentUserInfo!!.idBranch.toString())
            }
                .onSuccess {
                    try {
                        view!!.updateSpecialty(it.spec)
                    } catch (e: Exception) {
                        Timber.tag("my")
                            .e(getMessageForError(e, "DoctorsPresenter\$getSpecialtyByCenter$1 "))
                    }
                }.onFailure {
                    view!!.hideLoading()
                    Timber.tag("my").e(getMessageForError(it, "DoctorsPresenter\$getSpecialtyByCenter$2 "))
                    view!!.showErrorScreen()
                }
        }
    }

    fun getDoctorList(idSpec: Int) {
        if (idSpec == -1 && allDoc != null) {
            view!!.updateView(allDoc!!)
            view!!.hideLoading()
            return
        }
        if (allDoc != null) {
            view!!.updateView(sortAllDoc(idSpec))
            view!!.hideLoading()
            return
        }
        view!!.showLoading()

        mainScope.launch {
            kotlin.runCatching {
                networkManager.getAllDoctors(
                    prefManager.currentUserInfo!!.apiKey!!,
                    prefManager.centerInfo!!.db_name!!,
                    prefManager.currentUserInfo!!.idUser.toString(),
                    prefManager.currentUserInfo!!.idBranch.toString()
                )
            }
                .onSuccess {
                    if (it.mResponses.size <= 0) {
                        view!!.updateView(null)
                    } else {
                        allDoc = it.mResponses
                        if (view != null) view!!.updateView(it.mResponses)
                    }
                    view!!.hideLoading()
                }.onFailure {
                    Timber.tag("my").e(getMessageForError(it, "DoctorsPresenter\$getDoctorList "))
                    view!!.hideLoading()
                    view!!.showErrorScreen()
                }
        }
    }

    private fun sortAllDoc(idSpec: Int): List<AllDoctorsResponse> {
        val sortList: MutableList<AllDoctorsResponse> = ArrayList()
        for (i in allDoc!!.indices) {
            val doc = allDoc!![i]
            if (doc.getIdSpecialtiesIntList() == null) continue
            if (doc.getIdSpecialtiesIntList()!!.size == 1 && doc.getIdSpecialtiesIntList()!![0] == idSpec) {
                sortList.add(doc)
                continue
            }
            for (j in doc.getIdSpecialtiesIntList()!!.indices) {
                if (doc.getIdSpecialtiesIntList()!![j] == idSpec) {
                    sortList.add(doc)
                }
            }
        }
        return sortList
    }

    fun unSubscribe() {
        view!!.hideLoading()
    }

    fun getUserToken(): String {
        return prefManager.currentUserInfo!!.apiKey!!
    }

    fun getCurrentUser(): UserResponse {
        return prefManager.currentUserInfo!!
    }

    fun setCurrentUser(user: UserResponse) {
        prefManager.currentUserInfo = user
    }

}